scipy/linalg/_matfuncs_sqrtm.c

```diff
@@ -0,0 +1,1006 @@
+    // Get the number of slices to traverse if more than one; np.prod(shape[:-2])
+    npy_intp outer_size = 1;
+    if (ndim > 2)
+    {
+        for (int i = 0; i < ndim - 2; i++) { outer_size *= shape[i];}
+    }
+
+    // --------------------------------------------------------------------
+    // Workspace computation and allocation
+    // --------------------------------------------------------------------
+    int info = 0, sdim = 0, lwork = -1, intn = (int)n;
+    float tmp_float = 0.0f, one = 1.0f, zero = 0.0f;
+    sgees_("V", "N", NULL, &intn, NULL, &intn, &sdim, NULL, NULL, NULL, &intn, &tmp_float, &lwork, NULL, &info);
+    // Improbable to fail at lwork query but check anyway
+    if (info != 0) { *sq_info = -100; }
+    lwork = (int)tmp_float;
+    // 2n*n + 2n*n for data and vs for potentially converting to complex if needed
+    // n + n for wr, wi (needed for gees calls)
+    // lwork for work
+    size_t buffer_size = 4*n*n + 2*n + lwork;
+    float* restrict buffer = malloc(buffer_size*sizeof(float));
+    if (buffer == NULL) { *sq_info = -101; return; }
+
+    // --------------------------------------------------------------------
+    // Pointers to the variables inside the buffer for easy access
+    // --------------------------------------------------------------------
+    float* restrict data = &buffer[0];
+    float* restrict data2 = &buffer[n*n];
+    float* restrict vs = &buffer[2*n*n];
+    float* restrict wr = &buffer[4*n*n];
+    float* restrict wi = &buffer[4*n*n + n];
+    float* restrict work = &buffer[4*n*n + 2*n];
+    // Cover the first 2n*n and vs area with a complex pointer just in case
+    SCIPY_C* complex_data = &((SCIPY_C*)buffer)[0];
+    SCIPY_C* complex_vs = &((SCIPY_C*)buffer)[n*n];
+
+
+    /*====================================================================
+    |                    MAIN nxn SLICE LOOP                             |
+    ====================================================================*/
+    for (npy_intp idx = 0; idx < outer_size; idx++) {
+        /*------------------------------------------------------------------
+        Copy the current slice into buffer in F-layout for LAPACK calls. Note
+        that the input can be in C or F layout and not necessarily contiguous.
+        Since the input is predominantly C contiguous, we copy the data in
+        C-layout first and then transpose it to F-layout.
+
+        The output is always a C contiguous array hence we visit slices in
+        C-order for easy insertion to the output array; by just offsetting
+        idx*n*n * sizeof(dtype). For this reason, we  start the offset
+        calculation from the inner-most outer dimension.
+        ------------------------------------------------------------------*/
+        npy_intp offset = 0;
+        npy_intp temp_idx = idx;
+        for (int i = ndim - 3; i >= 0; i--) {
+            offset += (temp_idx % shape[i]) * strides[i];
+            temp_idx /= shape[i];
```

Comment:
Hopefully there's a check for zero-sized arrays at the call site. Probably in python. Needs a test.